src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -40,73 +42,52 @@ internal class AggregatorStore
+            double[] histogramBounds,
+            string[] tagKeysInteresting = null)
         {
             this.metrics = new MetricPoint[MaxMetricPoints];
             this.aggType = aggType;
             this.temporality = temporality;
             this.histogramBounds = histogramBounds;
             this.startTimeExclusive = DateTimeOffset.UtcNow;
-        }
-
-        internal int FindMetricAggregators(ReadOnlySpan<KeyValuePair<string, object>> tags)
-        {
-            int tagLength = tags.Length;
-            if (tagLength == 0)
-            {
-                this.InitializeZeroTagPointIfNotInitialized();
-                return 0;
-            }
-
-            var storage = ThreadStaticStorage.GetStorage();
-
-            storage.SplitToKeysAndValues(tags, tagLength, out var tagKey, out var tagValue);
-
-            if (tagLength > 1)
+            if (tagKeysInteresting == null)
             {
-                Array.Sort<string, object>(tagKey, tagValue);
+                this.updateLongCallback = this.UpdateLong;
+                this.updateDoubleCallback = this.UpdateDouble;
             }
-
-            return this.LookupAggregatorStore(tagKey, tagValue, tagLength);
-        }
-
-        internal void UpdateLong(long value, ReadOnlySpan<KeyValuePair<string, object>> tags)
-        {
-            try
+            else
             {
-                var index = this.FindMetricAggregators(tags);
-                if (index < 0)
+                this.updateLongCallback = this.UpdateLongCustomTags;
+                this.updateDoubleCallback = this.UpdateDoubleCustomTags;
+                var hs = new HashSet<string>(StringComparer.Ordinal);
+                foreach (var key in tagKeysInteresting)
                 {
-                    // TODO: Measurement dropped due to MemoryPoint cap hit.
-                    return;
+                    hs.Add(key);
                 }
 
-                this.metrics[index].Update(value);
-            }
-            catch (Exception)
-            {
-                // TODO: Measurement dropped due to internal exception.
+                this.tagKeysInteresting = hs;
+                this.tagsKeysInterestingCount = hs.Count;
             }
         }
 
-        internal void UpdateDouble(double value, ReadOnlySpan<KeyValuePair<string, object>> tags)
+        private delegate void UpdateLongDelegate(long value, ReadOnlySpan<KeyValuePair<string, object>> tags);
+
+        private delegate void UpdateDoubleDelegate(double value, ReadOnlySpan<KeyValuePair<string, object>> tags);
+
+        internal void Update(long value, ReadOnlySpan<KeyValuePair<string, object>> tags)
         {
-            try
-            {
-                var index = this.FindMetricAggregators(tags);
-                if (index < 0)
-                {
-                    // TODO: Measurement dropped due to MemoryPoint cap hit.
-                    return;
-                }
+            this.updateLongCallback(value, tags);
```

Comment:
How much time would this dispatch take?

Reply:
I tried a benchmark below and based on that, the difference between doing if check in hot path vs using delegate is very small. Delegate is slightly faster.

Below is .NetCore3.1

|                                Method |     Mean |    Error |   StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------------- |---------:|---------:|---------:|------:|------:|------:|----------:|
|             Delegate_ConditionCheckOnce | 17.71 ns | 0.481 ns | 1.358 ns |     - |     - |     - |         - |
| Direct_ConditionCheckOnEveryIteration | 17.30 ns | 0.370 ns | 0.426 ns |     - |     - |     - |         - |

Below is .Net5
|                                Method |     Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------------- |---------:|----------:|----------:|------:|------:|------:|----------:|
|             Delegate_ConditionCheckOnce| 4.225 ns | 0.1111 ns | 0.1279 ns |     - |     - |     - |         - |
| Direct_ConditionCheckOnEveryIteration | 6.410 ns | 0.1596 ns | 0.2018 ns |     - |     - |     - |         - |

and below is net462
|                                Method |     Mean |    Error |   StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------------- |---------:|---------:|---------:|------:|------:|------:|----------:|
|             Delegate_ConditionCheckOnce | 28.64 ns | 0.553 ns | 0.543 ns |     - |     - |     - |         - |
| Direct_ConditionCheckOnEveryIteration | 34.01 ns | 0.667 ns | 1.333 ns |     - |     - |     - |         - |


```
public class MetricGeneralBenchmarks
    {
        private UpdateDelegate updateDelegate;
        private Meter meter;
        private Counter<long> counter;
        private int randomNumber;

        private delegate void UpdateDelegate(long value);

        [GlobalSetup]
        public void Setup()
        {
            this.randomNumber = new Random().Next(1, 100);
            if (this.randomNumber % 5 == 0)
            {
                this.updateDelegate = this.UpdateA;
            }
            else
            {
                this.updateDelegate = this.UpdateB;
            }

            this.meter = new Meter("TestMeter");
            this.counter = this.meter.CreateCounter<long>("counter");
        }

        [Benchmark]
        public void Delegate_ConditionCheckOnce()
        {
            this.updateDelegate(100);
        }

        [Benchmark]
        public void Direct_ConditionCheckOnEveryIteration()
        {
            if (this.randomNumber % 5 == 0)
            {
                this.UpdateA(100);
            }
            else
            {
                this.UpdateB(100);
            }
        }

        private void UpdateA(long value)
        {
            this.counter.Add(value);
        }

        private void UpdateB(long value)
        {
            this.counter.Add(value);
        }
    }
```

Reply:
> I tried a benchmark below and based on that, the difference between doing if check in hot path vs using delegate is very small. Delegate is slightly faster.

This confirms that doing delegate+dispatch vs. doing a check in each loop is equally bad.
Sounds like both approaches won't work.

Maybe we should step back to something like https://github.com/open-telemetry/opentelemetry-dotnet/blob/bd189aa654948e79203d66d4b6d33fb93557d7e9/test/Benchmarks/Trace/TraceBenchmarks.cs#L114, while adding additional things (starting from no listener, to a listener which simply drops data on the floor, to a listener which passes data to processors, to a listener which passes data to a chain of processors), measuring what do we have to pay for.

I wrote a simple stress test to verify how bad an extra dispatch would be:

```csharp
public partial class Program
{
    private static readonly Meter TestMeter = new Meter("TestMeter", "1.0.0");
    private static readonly Counter<long> TestCounter = TestMeter.CreateCounter<long>("TestCounter");

    public static void Main()
    {
        var listener = new MeterListener();

        // listener.SetMeasurementEventCallback<long>(MeasurementEventCallback<long>);

        listener.SetMeasurementEventCallback<long>((instrument, value, tags, state) =>
        {
            // Console.WriteLine($"MeasurementEventCallback(\"{instrument.Name}\", {value}, {state})");
        });

        listener.EnableMeasurementEvents(TestCounter, null);
        Stress();
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    protected static void Run()
    {
        TestCounter.Add(100, new("key1", "value1"), new("key2", "value2"), new("key3", "value3"));
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    protected static void MeasurementEventCallback<T>(Instrument instrument, T value, ReadOnlySpan<KeyValuePair<string, object>> tags, object state)
    {
        // Console.WriteLine($"MeasurementEventCallback(\"{instrument.Name}\", {value}, {state})");
    }
}
```

Reply:
Stress test run results after ~200M loops (ran thrice)

main branch:
* Total Loops: 204,065,746
* Average Loops/Second: 4,635,856
* Average CPU Cycles/Loop: 3,399
----------------------------------------
* Total Loops: 202,408,337
* Average Loops/Second: 4,550,956
* Average CPU Cycles/Loop: 3,464
----------------------------------------
* Total Loops: 222,133,483
* Average Loops/Second: 4,283,082
* Average CPU Cycles/Loop: 3,639

this PR:
* Total Loops: 279,531,225
* Average Loops/Second: 4,566,384
* Average CPU Cycles/Loop: 3,449
----------------------------------------
* Total Loops: 205,510,447
* Average Loops/Second: 4,675,048
* Average CPU Cycles/Loop: 3,377
----------------------------------------
* Total Loops: 209,301,412
* Average Loops/Second: 4,778,571
* Average CPU Cycles/Loop: 3,278